### DISCUSSION TITLE
Create table with some cells highlighted  with ellipsis and with custom gridlines (custom gridlines  = table with custom row and colspans)

### DISCUSSION THREAD

**Discussion by kjambunathan** [kjambunathan/org-mode-ox-odt#267] [Show and tell]
[latex - Create table with circles around words and hiding some of the grid lines in Emacs Org-mode - Emacs Stack Exchange](https://emacs.stackexchange.com/questions/80418/create-table-with-circles-around-words-and-hiding-some-of-the-grid-lines-in-emac)

I am new to Emacs. Is there any way to create a table in Org mode like in the example table below. Particularly I\'d need to be able to circle some of the words and hide some of the grid lines. The end goal is in pdf format (ODT would also be useful).

**Example table:** ![](https://i.stack.imgur.com/f5MnJ.png) 

Owing to a very useful [link](https://stackoverflow.com/questions/10616460/merge-cells-in-org-mode-tables), this is about as far as I got:

``` example
+-----------+-----------------+------------------------------------------+
| Example #1|                                                            |
| *creative*|                                                            | 
+-----------+-----+-----------+------------------------------------------+
|           | A   | B         |                                          |
|           | A1  | B1        |           other bits                     |
|           | A1b | B1b       |                                          |
|           | A2  |   B2C2    |                                          |
+-----------+-----+-----------+------------------------------------------+
|   A2b  B2b C2b anjafkjnfjnef  janakjf  jknfajf b a jkf fa jbf a        |
| afjasf  afasf ajbf af  fajfasf a fa fasf af af  asf ajksf basf  as     |
+-----------+-----+-----------+------------------------------------------+
```

I have read that grid lines can be hidden [here](https://emacs.stackexchange.com/questions/59110/hide-table-border-lines-in-org-mode-tables). However, the suggested command `#+LATEX_HEADER: \renewcommand{\hline}{}`{.verbatim} gets rid of all horizontal grid lines when I need to get rid of all vertical grid lines as well as the intermediate horizontal grid lines instead. It might have something to do with cline, but I couldn\'t get this to work yet (see: [link](https://tex.stackexchange.com/questions/68980/removing-some-lines-in-table)).

The table, once complete, will form part of a larger Org file with plain text elements. Hence, I\'d like to get this done in Org mode. However, I am flexible in terms of the type of table used (org table, table.el etc.) as well as how to accomplish the circles and table editing (be it via Latex or other), only the output file should be in pdf format (or ODT). Does anyone have an idea?

**Comment by kjambunathan** [comment]
### zip file

[fancy-table-with-custom-grid-lines-and-highlights.zip](https://github.com/kjambunathan/org-mode-ox-odt/files/14391764/fancy-table-with-custom-grid-lines-and-highlights.zip)


To get this

![screenshot-2](https://github.com/kjambunathan/org-mode-ox-odt/assets/19869864/b68ffb13-ecef-4c77-a2dc-22f848a02709)




do this

```
#+odt_preferred_output_format: pdf

#+CAPTION: Table with Custom Grid Lines\\
#+CAPTION: Table with some Cells Highlighted with an "Elliptic" Box
#+ATTR_ODT: :span "@1$2{1:6}"
#+ATTR_ODT: :span "@2$1{4:1}"
#+ATTR_ODT: :span "@6$1{1:7}"
#+ATTR_ODT: :list-table t
#+ATTR_ODT: :style "GriddedTable"
- 
  - 
     Example #1

     Creative

  - 

  - 

  - 

  - 

  - 

  - 

- 

  - 

  - How Often? 

  - About once per month or less

  - 

    #+ATTR_ODT: :anchor "as-char" :shape "ellipse" :width 3  :style "OrgShape1"
    #+begin_customshape
         About once per week
    #+end_customshape

  - About each day

  - 2 or 3 times each day

  - More than 3 times each day

- 

  - 

  - How Intense?

  - Very Low

  - Low

  - Moderate

  - 

    #+ATTR_ODT: :anchor "as-char" :shape "ellipse" :width 3  :style "OrgShape1"
    #+begin_customshape
    High
    #+end_customshape

  - Very High

- 

  - 

  - How long lasting?

  - Less than 1 minute

  - 1-10 minute

  - 11-60 minutes

  - 

     #+ATTR_ODT: :anchor "as-char" :shape "ellipse" :width 3  :style "OrgShape1"
     #+begin_customshape
     1-4 hours
     #+end_customshape

  - Longer than 4 hours

- 

  - 

  - How easy to regulate?

  - Very easy?

  - 

     #+ATTR_ODT: :anchor "as-char" :shape "ellipse" :width 3  :style "OrgShape1"
     #+begin_customshape
     Easy
     #+end_customshape

  - Moderate

  - Difficult

  - Very  Difficult
  
- 

  - 

    In the example above, the person chose “About once per week” for
    “How Often” because they do not feel creative very frequently. The
    person chose “High” for ‘How Intense” because their creativity is
    strong when it does occur. The person chose “1-4 hours” for "How
    Long-Lasting” because when they feel creative the feeling lasts
    for a couple hours. Finally, the person chose “Easy” for ‘How Easy
    to Regulate” because when they feel creative they are able to
    increase this feeling by choosing appropriate activities.

  - 

  - 

  - 

  - 

  - 

  - 

    
This example uses [[https://github.com/kjambunathan/org-mode-ox-odt?tab=readme-ov-file#the-authoritative-fork-of-org-modes-odt-exporter][The Authoritative fork of Org mode's ODT exporter]]

- [[https://github.com/kjambunathan/org-mode-ox-odt/blob/fa8cadfe40175fef5a677c74b004e6848f7478b8/lisp/ox-odt.el#L3518][Custom Shapes]]
- [[https://github.com/kjambunathan/org-mode-ox-odt/blob/fa8cadfe40175fef5a677c74b004e6848f7478b8/lisp/ox-odt.el#L10387][List Tables]]
- [[https://github.com/kjambunathan/org-mode-ox-odt/blob/fa8cadfe40175fef5a677c74b004e6848f7478b8/lisp/ox-odt.el#L11470][Table with Row and Colspans]]
- [[https://github.com/kjambunathan/org-mode-ox-odt/blob/fa8cadfe40175fef5a677c74b004e6848f7478b8/lisp/ox-odt.el#L11834][Custom styles using nxml blocks]]


This document is produced with following version of ~ox-odt.el~

#+CAPTION: ~ox-odt~ Version Info
#+begin_example
commit fa8cadfe40175fef5a677c74b004e6848f7478b8 (HEAD -> master, origin/master, origin/HEAD)
Author: Jambunathan K <kjambunathan@gmail.com>
Date:   Sun Jan 21 15:59:10 2024 +0530
#+end_example

and rendered on following on version of ~LibreOffice~

#+CAPTION: ~Emacs~ Version Info
#+begin_example
Version: 7.6.4.1 (X86_64) / LibreOffice Community
Build ID: 60(Build:1)
CPU threads: 4; OS: Linux 6.5; UI render: default; VCL: gtk3
Locale: en-IN (en_IN); UI: en-US
Debian package version: 4:7.6.4~rc1-1+b2
Calc: threaded
#+end_example

The author uses following version of ~Emacs~

#+CAPTION: ~Emacs~ Version Info
#+begin_example
$ emacs --version

GNU Emacs 30.0.50
Development version 5af4e346b0b0 on master branch; build date 2024-02-09.
Copyright (C) 2024 Free Software Foundation, Inc.
GNU Emacs comes with ABSOLUTELY NO WARRANTY.
You may redistribute copies of GNU Emacs
under the terms of the GNU General Public License.
For more information about these matters, see the file named COPYING.
#+end_example


#+ATTR_ODT: :target "master_styles" :backends (odt)
#+begin_src nxml
<style:master-page style:name="Standard"
                   style:page-layout-name="A4LandscapeLayout">
  <style:footer>
    <text:p text:style-name="OrgFooter">
      <text:page-number text:select-page="current" />
    </text:p>
  </style:footer>
</style:master-page>
#+end_src


#+ATTR_ODT: :target "automatic_styles" :backends (odt)
#+begin_src nxml
<style:style style:name="OrgShape1"
             style:family="graphic">
  <style:graphic-properties draw:auto-grow-height="true"
                            draw:fill-color="#ffffff"
                            draw:textarea-horizontal-align="justify"
                            draw:textarea-vertical-align="middle"
                            draw:wrap-influence-on-position="once-concurrent"
                            fo:margin-bottom="0cm"
                            fo:margin-left="0cm"
                            fo:margin-right="0cm"
                            fo:margin-top="0cm"
                            fo:min-height="0cm"
                            fo:min-width="0cm"
                            style:flow-with-text="true"
                            style:horizontal-pos="from-left"
                            style:horizontal-rel="paragraph"
                            style:repeat="no-repeat"
                            style:run-through="foreground"
                            style:vertical-pos="from-top"
                            style:vertical-rel="paragraph"
                            style:wrap="none"
                            style:wrap-contour-mode="full"
                            svg:stroke-color="#000000"
                            svg:stroke-width="0cm" />
  <style:paragraph-properties style:writing-mode="lr-tb" />
</style:style>
#+end_src

```

**Comment by QiangF** [comment]
I think one possible improvement for list table is if a row of a multi-row table is missing item, it should be left empty without complaint/error, the column number should be derived from the row with the maximum items, then we don't need those extra marks.

**Comment by tiberius-starvoyager** [comment]
Following the discussion [here](https://emacs.stackexchange.com/questions/80418/create-table-with-circles-around-words-and-hiding-some-of-the-grid-lines-in-emac/80489?noredirect=1#comment135259_80489) I tried install your exporter using the steps recommended by you. After invoking the initial steps `M-x customize-variable` RET  `package-archives` RET and adding your archive (kjambunathan.github.io/elpa), I ran `M-x package-refresh-contents`  RET and `M-x package-install` with `ox-odt`. Upon restarting Emacs, I  encountered an error, however: `emacs Invalid function:  org-assert-version`. In the `ox-odt-compile-log` there were several warnings, saying that the `org-assert-version` was not known to be defined. I then restarted Emacs using, `emacs --debug-init`. Somehow this seems to have done the trick. Not only am I back to my  configuration, but I can also access ox-odt.el and your script works beautifully!! I just needed to add the `:height` parameter after `:width 3` as the ellipses were too small on my system. I am not sure though whether `--debug-init` has sorted the initial problem for good or if there is something else I might be missing. Do you happen to have any experience with this? In any case, thank you for your support up to now. This has been phenomenal!

**Reply by QiangF** [reply]
It doesn't need to be that complicated! Just 2 steps.

     1. remove references to built-in org in path
     (setq load-path (cl-remove-if (lambda (x) (string-match-p "org$" x)) load-path))


     2. clone this repo, add the lisp dir to your load path (https://github.com/kjambunathan/org-mode-ox-odt/tree/master/lisp)

**Reply by kjambunathan** [reply]
I really doubt if  `emacs --debug-init` fixed the issues for you.  There is something else happening your back (and I don't see what it is).  Since you have already been able to load the **enhanced** `ox-odt` , all you have to do is to fix the `load-path` so that `ox-odt` comes from the ELPA path and NOT FROM Emacs or Org-mode code.

_(I wrote the below note for a future user who might run in to issues with `org-assert-version`)_

### Back story behind `org-assert-version`

While installing `ox-odt` from https://kjambunathan.github.io/elpa/,  it first installs `org-9.6`.  This is because [`org-9.6` is a dependency](https://github.com/kjambunathan/kjambunathan.github.io/blob/44aa5a302eb518e811865370fee119d4578d4b4c/elpa/archive-contents#L8). Once the builtin `org.el` is upgraded to `org-9.6`, it will subsequently install MY `ox-odt.el`.

`ox-odt.el` that comes with Emacs or Vanilla Org mode [DOES USE `org-assert-version`](https://git.savannah.gnu.org/cgit/emacs.git/tree/lisp/org/ox-odt.el?h=5af4e346b0b078d6e8f3dd90bb66899d3ed99810#n29)

But the `ox-odt.el` that comes with this fork [DOES NOT USE `org-assert-version`](https://github.com/kjambunathan/org-mode-ox-odt/blob/579069cb28400be6dee60d6cc08e6a4664a015db/lisp/ox-odt.el#L31)

So, the compilation error you are seeing wrt `ox-odt.el` is the error that happens when `ox-odt.el` from stock GNU ELPA version of `org-9.6` is compiled.

To work around the `org-assert-version` issues,  see [Irreal: Update On The Org-9.6 Problem](https://irreal.org/blog/?p=10999).  I am reproducing the steps outlined in that post,

### How to fix `org-assert-vesion` issues

The following instructions assume that `package-user-dir` is at its
default value of `~/.emacs.d/elpa/`.  

*(If the value is a non-standard one, adjust the paths below suitably)*

- **Uninstall `org` installed by GNU ELPA**

   - `M-x list-packages`
   - Go to line `org` line (it should say `installed`)
   - do `d x` (=`delete`)

- **Uninstall `ox-odt` installed from `kjambunathan.github.io`**

   - `M-x list-packages`
   - Go to line `ox-odt` line (it should say `installed`)
   - do `d x` (=`delete`)

- **Make sure there is no `org-*` directory in ELPA path**

   `C-x C-f ~/.emacs.d/elpa/` should not have ANY  `org-*` directory
   
- **Quit Emacs**

  `C-x C-c`

- **Start Emacs Minimally**

  `emacs -Q` (= `BIG/CAPITAL Q`)

- **Install `org` from GNU ELPA**

  - `M-x list-packages`
  - Go to line that says `org`
  - Do `i x` (=`install`)

- **Quit Emacs**

  `C-x C-c`

- **Restart Emacs normally**

  `emacs`

- **Check the version of `org` you  are running**

  `M-x org-version`

- **Install my `ox-odt`**

  - `M-x list-packages`
  - Go to line that says `ox-odt` (it should mention my ELPA path)
  - Do `i x`

Above steps will ensure that you have a fairly clean installation.

### Adjust the `load-path` so that `ox-odt` comes from `kjambunathan.github.io`

- **Restart Emacs**

- **Check the directory from which `org` comes from**: `M-x find-library org`, `C-x d`

- **Check the directory from which `ox-odt` comes from**: `M-x find-library ox-odt`, `C-x d`

    If this directory is **NOT `~/.emacs.d/elpa/ox-odt-<whatever>`** you
    may have to fix the `load-path` by adding the following snippet to your `init.el`.

     ```elisp
	 (progn
	   ;; (add-to-list 'load-path "~/.emacs.d/elpa/ox-odt-9.6.620/")

	   (when (featurep 'ox-odt)
		 (message "ox-odt is already loaded from %s.  Move this snippet to the beginning of `init.el`."
			  (find-library "ox-odt")))

	   ;; Use the highest version of `ox-odt' from `package-user-dir'
	   (add-to-list 'load-path
				(let ((default-directory "~/.emacs.d/elpa/"))
			  (car (last (sort (file-expand-wildcards "ox-odt-*" 'full) 'string> )))))
	   (eval-after-load 'org '(require 'ox-odt nil nil)))
     ```

### Small utility to find who loads a package of interest behind your back


Add this snippet your to the very beginning of `init.el` so that you
can catch if `org`, or `ox-odt` gets loaded behind your back
(presumaby by some `use-package` config you haven't written yourself)

```elisp
(setq debug-on-error t)

(defmacro debug-on--load (feature)
  `(with-eval-after-load ',feature
     (message "Loaded `%s': %s" ',feature
          (find-library-name (symbol-name ',feature)))
     ;; Comment the `(debug)' out may be.
     ;; If it is ON, it will give you a hint how the `feature' gets
     ;; loaded behind your back.  Type `c`, when Emacs enters debugger.
     (debug)))

(defmacro debug-on-load (feature-or-features)
  (let* ((features (or (unless (consp feature-or-features)
             (list feature-or-features))
               feature-or-features)))
    `(progn
       ,@(seq-map
      (lambda (it)
        `(debug-on--load ,it))
      features))))

(debug-on-load
 (
  ;; Add libraries of interest here
  org
  ox-odt
  ))
```

**Reply by kjambunathan** [reply]
>  I just needed to add the `:height` parameter after `:width 3` as the ellipses were too small on my system. 

It is enough to specify adjust the width (= No need to specify height).  See https://github.com/kjambunathan/org-mode-ox-odt/issues/268

**Comment by kjambunathan** [comment]
> It doesn't need to be that complicated! Just 2 steps.
> 
> ```
>  1. remove references to built-in org in path
>  (setq load-path (cl-remove-if (lambda (x) (string-match-p "org$" x)) load-path))
> 
> 
>  2. clone this repo, add the lisp dir to your load path (https://github.com/kjambunathan/org-mode-ox-odt/tree/master/lisp)
> ```

After cloning the repo, do a `make` so that stuff gets compiled.

@tiberius-starvoyager, if you are using the `git clone` of this directory,  then you may have to [install `peg`, `seq`, and `map` etc.](https://github.com/kjambunathan/org-mode-ox-odt/discussions/133#discussioncomment-8529785)  manually.   `git clone` will give you a very up-to-date exporter, specifically a ODS/XLSX exporter.   The disadvantage is it could be a hassle to setup (YMMV).  

The `ox-odt` you installed from kjambunathan.github.io is good enough for all practical purposes, even though it could be an year old.  

Whatever you choose `git clone` or my ELPA packages, if you run in to further  issues I will be happy to help.
